credentials/endpointcreds/internal/client/client.go

```diff
@@ -0,0 +1,147 @@
+	Retryer retry.Retryer
+}
+
+// Copy creates a copy of the API options.
+func (o Options) Copy() Options {
+	to := o
+	to.APIOptions = append([]func(*smithymiddleware.Stack) error{}, o.APIOptions...)
+	return to
+}
+
+// Client is an client for retrieving AWS credentials from an endpoint
+type Client struct {
+	options Options
+}
+
+// New constructs a new Client from the given options
+func New(options Options, optFns ...func(*Options)) *Client {
+	options = options.Copy()
+
+	if options.HTTPClient == nil {
+		options.HTTPClient = aws.NewBuildableHTTPClient().WithTransportOptions(func(t *http.Transport) {
+			t.Proxy = nil
+		})
+	}
+
+	if options.Retryer == nil {
+		options.Retryer = retry.NewStandard()
+	}
+
+	for _, fn := range optFns {
+		fn(&options)
+	}
+
+	client := &Client{
+		options: options,
+	}
+
+	return client
+}
+
+// NewFromConfig constructs a new client using the provided SDK config.
+func NewFromConfig(cfg aws.Config, optsFns ...func(options *Options)) *Client {
+	opts := Options{
+		HTTPClient: cfg.HTTPClient,
+		Retryer:    cfg.Retryer,
+	}
+
+	return New(opts, optsFns...)
+}
+
+// GetCredentials retrieves credentials from credential endpoint
+func (c *Client) GetCredentials(ctx context.Context, optFns ...func(*Options)) (*GetCredentialsOutput, error) {
+	stack := smithymiddleware.NewStack("GetCredentials", smithyhttp.NewStackRequest)
+	options := c.options.Copy()
+	for _, fn := range optFns {
+		fn(&options)
+	}
+
+	stack.Serialize.Add(&serializeOpGetCredential{AuthorizationToken: options.AuthorizationToken}, smithymiddleware.After)
+	stack.Build.Add(&buildEndpoint{Endpoint: options.Endpoint}, smithymiddleware.After)
+	stack.Deserialize.Add(&deserializeOpGetCredential{}, smithymiddleware.After)
+
+	for _, fn := range options.APIOptions {
+		fn(stack)
+	}
+
+	handler := smithymiddleware.DecorateHandler(smithyhttp.NewClientHandler(options.HTTPClient), stack)
+	result, _, err := handler.Handle(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return result.(*GetCredentialsOutput), err
+}
+
+// GetCredentialsOutput is the response from the credential endpoint
+type GetCredentialsOutput struct {
+	Expiration      *time.Time
+	AccessKeyID     string
+	SecretAccessKey string
+	Token           string
+}
+
+// EndpointError is an error returned from the endpoint service
+type EndpointError struct {
+	Code    string `json:"code"`
+	Message string `json:"message"`
+}
+
+// Error is the error mesage string
+func (e *EndpointError) Error() string {
+	return fmt.Sprintf("%s: %s", e.Code, e.Message)
+}
+
+// ErrorCode is the error code returned by the endpoint
+func (e *EndpointError) ErrorCode() string {
+	return e.Code
+}
+
+// ErrorMessage is the error message returned by the endpoint
+func (e *EndpointError) ErrorMessage() string {
+	return e.Message
+}
+
+// ErrorFault indicates error fault classification
+func (e *EndpointError) ErrorFault() smithy.ErrorFault {
+	return smithy.FaultUnknown
```

Comment:
Seems odd that this will always be Unknown fault.

Reply:
guessing based on HTTP status code i think is only alternative.

Reply:
Can update it to base the fault on client / server based on HTTP Status code